In [1]:
import random
import numpy as np
from pathlib import Path
from the_well.data import WellDataset

import torch
from torch.utils.data import DataLoader

from modules import *
from losses import *
from datasets import HelmholtzDataset

In [2]:
SEED         = 42
EPOCHS       = 20
BATCH_SIZE   = 8
LR           = 1e-3

DATASET_PATH = "/mnt/storage_C1/BILL_pino/the_well/datasets/helmholtz_staircase"
OUTPUT_DIR   = "./models"

MODES1       = 16   # Modos de Fourier na primeira dimensão espacial (x)
MODES2       = 16   # Modos de Fourier na segunda dimensão espacial (y)
WIDTH        = 32   # Número de canais internos (largura do modelo)
DEPTH        = 4    # Quantidade de camadas de Fourier
PROJ_DIM     = 128  # Dimensão da MLP de projeção para a saída

In [3]:
random.seed(SEED)
np.random.seed(SEED)
torch.manual_seed(SEED)
torch.cuda.manual_seed_all(SEED)

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print(f"Dispositivo: {device}")
print(f"Torch: {torch.__version__}  |  Torchvision: {__import__('torchvision').__version__}")

Dispositivo: cuda
Torch: 2.11.0+cu128  |  Torchvision: 0.26.0+cu128


In [4]:
train_dataset = WellDataset(
    path=DATASET_PATH,
    well_split_name="train"
)

validation_dataset = WellDataset(
    path=DATASET_PATH,
    well_split_name="valid"
)

train_ds = HelmholtzDataset(train_dataset)
val_ds = HelmholtzDataset(validation_dataset)

train_loader = DataLoader(
    train_ds,
    batch_size=BATCH_SIZE,
    shuffle=True,
    num_workers=4,
    pin_memory=True,
    persistent_workers=True,
    prefetch_factor=4,
)

val_loader = DataLoader(
    val_ds,
    batch_size=BATCH_SIZE,
    shuffle=False,
    num_workers=2,
    pin_memory=True,
    persistent_workers=False,
    prefetch_factor=4,
)

### Teste 1

In [5]:
x, y = train_ds[0]
in_dim, out_dim = x.shape[-1], y.shape[-1] 

model = FNO2d(
    modes1=MODES1,
    modes2=MODES2,
    width=WIDTH,
    in_dim=in_dim,
    out_dim=out_dim,
    depth=DEPTH,
    proj_dim=PROJ_DIM
).cuda()
model = torch.compile(model)

optimizer = torch.optim.AdamW(
    model.parameters(),
    lr=LR
)

criterion = relative_l2_loss

/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


In [6]:
history = train_model(
    model=model,
    train_loader=train_loader,
    val_loader=val_loader,
    optimizer=optimizer,
    criterion=criterion,
    epochs=EPOCHS,
    checkpoint_dir=OUTPUT_DIR
)

Epoch 1/20:   0%|          | 0/2548 [00:00<?, ?it/s]/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/torch/_inductor/lowering.py:2212: UserWarning: Torchinductor does not support code generation for complex operators. Performance may be worse than eager.
  warnings.warn(
W0622 19:51:55.923000 17941 torch/_inductor/utils.py:1731] [0/0] Not enough SMs to use max_autotune_gemm mode
Epoch 1/20: 100%|██████████| 2548/2548 [21:37<00:00,  1.96it/s, loss=0.991098]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `p

Epoch 001 | train = 0.991098 | val = 0.845437 | best_val = 0.845437


Epoch 2/20: 100%|██████████| 2548/2548 [20:14<00:00,  2.10it/s, loss=0.987134]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 002 | train = 0.987134 | val = 0.807976 | best_val = 0.807976


Epoch 3/20: 100%|██████████| 2548/2548 [20:02<00:00,  2.12it/s, loss=0.987013]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 003 | train = 0.987013 | val = 0.778811 | best_val = 0.778811


Epoch 4/20: 100%|██████████| 2548/2548 [17:40<00:00,  2.40it/s, loss=0.986145]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 004 | train = 0.986145 | val = 0.820198 | best_val = 0.778811


Epoch 5/20: 100%|██████████| 2548/2548 [13:08<00:00,  3.23it/s, loss=0.986210]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 005 | train = 0.986210 | val = 0.812756 | best_val = 0.778811


Epoch 6/20: 100%|██████████| 2548/2548 [10:57<00:00,  3.87it/s, loss=0.986154]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 006 | train = 0.986154 | val = 0.793645 | best_val = 0.778811


Epoch 7/20: 100%|██████████| 2548/2548 [09:52<00:00,  4.30it/s, loss=0.986439]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 007 | train = 0.986439 | val = 0.790209 | best_val = 0.778811


Epoch 8/20: 100%|██████████| 2548/2548 [09:54<00:00,  4.28it/s, loss=0.985601]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 008 | train = 0.985601 | val = 0.784678 | best_val = 0.778811


Epoch 9/20: 100%|██████████| 2548/2548 [09:52<00:00,  4.30it/s, loss=0.986033]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 009 | train = 0.986033 | val = 0.758614 | best_val = 0.758614


Epoch 10/20: 100%|██████████| 2548/2548 [09:45<00:00,  4.35it/s, loss=0.985596]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 010 | train = 0.985596 | val = 0.841660 | best_val = 0.758614


Epoch 11/20: 100%|██████████| 2548/2548 [09:47<00:00,  4.34it/s, loss=0.985960]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 011 | train = 0.985960 | val = 0.798724 | best_val = 0.758614


Epoch 12/20: 100%|██████████| 2548/2548 [10:44<00:00,  3.96it/s, loss=0.985351]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 012 | train = 0.985351 | val = 0.808206 | best_val = 0.758614


Epoch 13/20: 100%|██████████| 2548/2548 [10:15<00:00,  4.14it/s, loss=0.985129]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 013 | train = 0.985129 | val = 0.806658 | best_val = 0.758614


Epoch 14/20: 100%|██████████| 2548/2548 [10:26<00:00,  4.06it/s, loss=0.985139]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 014 | train = 0.985139 | val = 0.795372 | best_val = 0.758614


Epoch 15/20: 100%|██████████| 2548/2548 [10:14<00:00,  4.15it/s, loss=0.984761]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 015 | train = 0.984761 | val = 0.798778 | best_val = 0.758614


Epoch 16/20: 100%|██████████| 2548/2548 [10:06<00:00,  4.20it/s, loss=0.984801]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 016 | train = 0.984801 | val = 0.784086 | best_val = 0.758614


Epoch 17/20: 100%|██████████| 2548/2548 [09:46<00:00,  4.34it/s, loss=0.984828]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 017 | train = 0.984828 | val = 0.790830 | best_val = 0.758614


Epoch 18/20: 100%|██████████| 2548/2548 [09:56<00:00,  4.27it/s, loss=0.985518]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 018 | train = 0.985518 | val = 0.792563 | best_val = 0.758614


Epoch 19/20: 100%|██████████| 2548/2548 [09:56<00:00,  4.27it/s, loss=0.985248]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 019 | train = 0.985248 | val = 0.790036 | best_val = 0.758614


Epoch 20/20: 100%|██████████| 2548/2548 [09:53<00:00,  4.29it/s, loss=0.985060]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 020 | train = 0.985060 | val = 0.788603 | best_val = 0.758614


Resultado teste 1

Epoch 020 | train = 0.985060 | val = 0.788603 | best_val = 0.758614

### Teste 2

In [7]:
x, y = train_ds[0]
in_dim, out_dim = x.shape[-1], y.shape[-1] 

model = FNO2d_v1(
    modes1=MODES1,
    modes2=MODES2,
    width=64,
    width1= 64,
    in_dim=in_dim,
    out_dim=out_dim,
    depth=DEPTH,
    proj_dim=192
).cuda()
model = torch.compile(model)

optimizer = torch.optim.AdamW(
    model.parameters(),
    lr=LR
)

criterion = relative_l2_loss

In [8]:
history2 = train_model(
    model=model,
    train_loader=train_loader,
    val_loader=val_loader,
    optimizer=optimizer,
    criterion=criterion,
    epochs=EPOCHS,
    checkpoint_dir=OUTPUT_DIR
)

Epoch 1/20: 100%|██████████| 2548/2548 [18:47<00:00,  2.26it/s, loss=0.994959]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 001 | train = 0.994959 | val = 0.847745 | best_val = 0.847745


Epoch 2/20: 100%|██████████| 2548/2548 [18:38<00:00,  2.28it/s, loss=0.988145]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 002 | train = 0.988145 | val = 0.781171 | best_val = 0.781171


Epoch 3/20: 100%|██████████| 2548/2548 [18:38<00:00,  2.28it/s, loss=0.987868]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 003 | train = 0.987868 | val = 0.831746 | best_val = 0.781171


Epoch 4/20: 100%|██████████| 2548/2548 [18:38<00:00,  2.28it/s, loss=0.986697]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 004 | train = 0.986697 | val = 0.805427 | best_val = 0.781171


Epoch 5/20: 100%|██████████| 2548/2548 [18:38<00:00,  2.28it/s, loss=0.986328]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 005 | train = 0.986328 | val = 0.795265 | best_val = 0.781171


Epoch 6/20: 100%|██████████| 2548/2548 [18:38<00:00,  2.28it/s, loss=0.986580]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 006 | train = 0.986580 | val = 0.811510 | best_val = 0.781171


Epoch 7/20: 100%|██████████| 2548/2548 [18:38<00:00,  2.28it/s, loss=0.986410]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 007 | train = 0.986410 | val = 0.813498 | best_val = 0.781171


Epoch 8/20: 100%|██████████| 2548/2548 [18:38<00:00,  2.28it/s, loss=0.985853]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 008 | train = 0.985853 | val = 0.796689 | best_val = 0.781171


Epoch 9/20: 100%|██████████| 2548/2548 [18:38<00:00,  2.28it/s, loss=0.986052]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 009 | train = 0.986052 | val = 0.807781 | best_val = 0.781171


Epoch 10/20: 100%|██████████| 2548/2548 [18:38<00:00,  2.28it/s, loss=0.985728]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 010 | train = 0.985728 | val = 0.798156 | best_val = 0.781171


Epoch 11/20: 100%|██████████| 2548/2548 [18:38<00:00,  2.28it/s, loss=0.985824]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 011 | train = 0.985824 | val = 0.808782 | best_val = 0.781171


Epoch 12/20: 100%|██████████| 2548/2548 [18:38<00:00,  2.28it/s, loss=0.985808]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 012 | train = 0.985808 | val = 0.793910 | best_val = 0.781171


Epoch 13/20: 100%|██████████| 2548/2548 [18:38<00:00,  2.28it/s, loss=0.985987]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 013 | train = 0.985987 | val = 0.798869 | best_val = 0.781171


Epoch 14/20: 100%|██████████| 2548/2548 [18:38<00:00,  2.28it/s, loss=0.985716]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 014 | train = 0.985716 | val = 0.804609 | best_val = 0.781171


Epoch 15/20: 100%|██████████| 2548/2548 [18:38<00:00,  2.28it/s, loss=0.985319]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 015 | train = 0.985319 | val = 0.789468 | best_val = 0.781171


Epoch 16/20: 100%|██████████| 2548/2548 [18:38<00:00,  2.28it/s, loss=0.985173]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 016 | train = 0.985173 | val = 0.796989 | best_val = 0.781171


Epoch 17/20: 100%|██████████| 2548/2548 [18:38<00:00,  2.28it/s, loss=0.985189]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 017 | train = 0.985189 | val = 0.790856 | best_val = 0.781171


Epoch 18/20: 100%|██████████| 2548/2548 [18:38<00:00,  2.28it/s, loss=0.984662]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 018 | train = 0.984662 | val = 0.781061 | best_val = 0.781061


Epoch 19/20: 100%|██████████| 2548/2548 [18:38<00:00,  2.28it/s, loss=0.985519]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 019 | train = 0.985519 | val = 0.793971 | best_val = 0.781061


Epoch 20/20: 100%|██████████| 2548/2548 [18:38<00:00,  2.28it/s, loss=0.984769]
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(
/home/al.bruno.paula/simul/env/lib/python3.12/site-packages/the_well/data/datasets.py:736: RuntimeWarning: Only axis-aligned boundary fully supported. Boundary for axis counted as `open` or `periodic` if any part of it is and `wall` otherwise.If this does not fit your desired usecase, set `boundary_return_type=None`.
  warnings.warn(


Epoch 020 | train = 0.984769 | val = 0.789566 | best_val = 0.781061


Resultados Teste 2

Epoch 020 | train = 0.984769 | val = 0.789566 | best_val = 0.781061

### Teste 3

In [9]:
x, y = train_ds[0]
in_dim, out_dim = x.shape[-1], y.shape[-1] 

model = FNO2d_v1(
    modes1=MODES1,
    modes2=MODES2,
    width=WIDTH,
    width1= WIDTH,
    in_dim=in_dim,
    out_dim=out_dim,
    depth=DEPTH,
    proj_dim=PROJ_DIM
).cuda()
model = torch.compile(model)

optimizer = torch.optim.AdamW(
    model.parameters(),
    lr=LR
)

criterion = relative_l2_loss

In [ ]:
history3 = train_model(
    model=model,
    train_loader=train_loader,
    val_loader=val_loader,
    optimizer=optimizer,
    criterion=criterion,
    epochs=EPOCHS,
    checkpoint_dir=OUTPUT_DIR
)

Epoch 1/20:   1%|          | 29/2548 [00:14<10:32,  3.98it/s, loss=1.050622] 